sys/debug/virtmem.py

```diff
@@ -1,43 +1,207 @@
-import gdb
-
 from .struct import TailQueue
-from .cmd import UserCommand
+from .cmd import UserCommand, CommandDispatcher
 from .cpu import TLBLo
-from .utils import TextTable, global_var, cast
+from .utils import TextTable, global_var, cast_ptr, get_arch
+from .proc import Process
 
 
 PM_NQUEUES = 16
 
 
-class PhysMap(UserCommand):
+class VmInfo(CommandDispatcher):
+    """Examine virtual memory data structures."""
+
+    def __init__(self):
+        super().__init__('vm', [KernelPmap(),
+                                UserPmap(),
+                                VmMapDump(),
+                                CurrentSegmentInfo(),
+                                ProcSegmentInfo(),
+                                VmPhysSeg(),
+                                VmFreePages(),
+                                ])
+
+
+def _print_mips_pmap(pmap):
```

Comment:
Please move all machine dependant procedures to `cpu.py`. There provide a separate class for each CPU.

After it's done export this set of classes that are valid for the current architecture. Then provide class / function aliases at the top level like so:

```python
class PageTableAArch64:
   ...
   
def PrintPageTableAarch64(...):
   ...

if get_arch() == 'aarch64':
   PageTable = PageTableAArch64
   PrintPageTable = PrintPageTableAArch64
elif get_arch() == 'mips':
   PageTable = PageTableMIPS
   PrintPageTable = PrintPageTableMIPS
else:
   ...
```

Reply:
I have done something like that.

Reply:
For now only Mips is supported, but I have provided stup classes for other architectures.